app/v1/controllers/validators/CreateAndAmendOtherDeductionsValidatorFactory.scala

```diff
@@ -0,0 +1,157 @@
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import cats.implicits._
+import play.api.libs.json.JsValue
+import v1.models.request.createAndAmendOtherDeductions.{CreateAndAmendOtherDeductionsBody, CreateAndAmendOtherDeductionsRequestData, Seafarers}
+
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+import scala.util.{Failure, Success, Try}
+
+class CreateAndAmendOtherDeductionsValidatorFactory {
+
+  private val resolveJson = new ResolveJsonObject[CreateAndAmendOtherDeductionsBody]()
+
+  def validator(nino: String, taxYear: String, body: JsValue): Validator[CreateAndAmendOtherDeductionsRequestData] =
+    new Validator[CreateAndAmendOtherDeductionsRequestData] {
+
+      def validate: Validated[Seq[MtdError], CreateAndAmendOtherDeductionsRequestData] =
+        (
+          ResolveNino(nino),
+          ResolveTaxYear(taxYear),
+          resolveJson(body)
+        ).mapN(CreateAndAmendOtherDeductionsRequestData) andThen validateBody
+
+      private def validateBody(
+          parsed: CreateAndAmendOtherDeductionsRequestData): Validated[Seq[MtdError], CreateAndAmendOtherDeductionsRequestData] = {
+        import parsed.body._
+        List(
+          validateBodyFieldFormat(seafarers),
+          validateBodyDateRange(seafarers)
+        ).traverse(identity)
+          .map(_ => parsed)
+      }
+
+    }
+
+  private def validateBodyFieldFormat(seafarers: Option[Seq[Seafarers]]): Validated[Seq[MtdError], Unit] = {
+    val validationResult = seafarers
+      .getOrElse(Seq.empty[Seafarers])
+      .zipWithIndex
+      .traverse_ { case (item, index) => validateSeafarers(item, index) }
+
+    validationResult
+  }
+
+  private def validateBodyDateRange(seafarers: Option[Seq[Seafarers]]): Validated[Seq[MtdError], Unit] = {
+    val validationResult = seafarers
+      .getOrElse(Seq.empty[Seafarers])
+      .zipWithIndex
+      .traverse_ { case (item, index) => validateToDateBeforeFromDate(item, index) }
+
+    validationResult
+  }
+
+  private def validateSeafarers(seafarers: Seafarers, arrayIndex: Int): Validated[Seq[MtdError], Unit] = {
+    (
+      ValidateCustomerReference.validateOptional(field = seafarers.customerReference, path = s"/seafarers/$arrayIndex/customerReference"),
+      validateAmountDeducted(value = seafarers.amountDeducted, path = s"/seafarers/$arrayIndex/amountDeducted"),
+      validateNameOfShip(
+        field = seafarers.nameOfShip,
+        path = s"/seafarers/$arrayIndex/nameOfShip"
+      ),
+      ValidateDate.validateDateFormat(
+        field = seafarers.fromDate,
+        path = s"/seafarers/$arrayIndex/fromDate"
+      ),
+      ValidateDate.validateDateFormat(
+        field = seafarers.toDate,
+        path = s"/seafarers/$arrayIndex/toDate"
+      )
+    ).tupled
+      .andThen { case (_, _, _, _, _) => Validated.Valid(()) }
+  }
+
+  object ValidateCustomerReference {
+    val customerRefRegex: String = "^[0-9a-zA-Z{À-˿’}\\- _&`():.'^]{1,90}$"
```

Comment:
shouldn't specify the string type, + can be a private val

Reply:
Maybe worth turning it into a regex type (add .r to the end), & swap the "matches" around, otherwise the Regex instance gets constructed every time it's used